app/src/main/java/com/duckduckgo/app/systemsearch/DeviceAppSuggestionsAdapter.kt

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright (c) 2020 DuckDuckGo
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.duckduckgo.app.systemsearch
+
+import android.view.LayoutInflater
+import android.view.View
+import android.view.ViewGroup
+import android.widget.ImageView
+import android.widget.TextView
+import androidx.annotation.UiThread
+import androidx.recyclerview.widget.RecyclerView
+import com.duckduckgo.app.browser.R
+import com.duckduckgo.app.systemsearch.DeviceAppSuggestionsAdapter.DeviceAppViewHolder
+import kotlinx.android.synthetic.main.item_device_app_suggestion.view.*
+
+class DeviceAppSuggestionsAdapter(
+    private val clickListener: (DeviceApp) -> Unit
+) : RecyclerView.Adapter<DeviceAppViewHolder>() {
+    private var deviceApps: List<DeviceApp> = ArrayList()
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): DeviceAppViewHolder {
+        val root = LayoutInflater.from(parent.context).inflate(R.layout.item_device_app_suggestion, parent, false)
+        return DeviceAppViewHolder(root, root.icon, root.title)
+    }
+
+    override fun onBindViewHolder(holder: DeviceAppViewHolder, position: Int) {
+        val app = deviceApps[position]
+        holder.title.text = app.shortName
+        holder.root.setOnClickListener {
+            clickListener(app)
+        }
+        val drawable = holder.icon.context.packageManager.getApplicationIcon(app.packageName)
```

Comment:
should the adapter call the packageManger every time here? Maybe this could be already in `DeviceApp` or the lookup done at the `ViewModel` level?

Reply:
Doing it upfront for all DeviceApp instances is expensive and noticeably slows down initial device app lookups so I've opted to do the work here as needed on a small number of items at a time. To your point we're though we're making this call multiple times if a user repeats their search terms so I'm moving the storage of the icon into DeviceApps, but only storing these lazily. This avoids the upfront cost of a mass lookup and removes duplicate lookups. Thanks!